packages/web-wallet/src/contexts/hooks/useWalletConnection.ts

```diff
@@ -0,0 +1,502 @@
+      if (typeof xpubResponse === 'string') {
+        try {
+          parsedResponse = JSON.parse(xpubResponse);
+        } catch (parseError) {
+          console.error('Failed to parse xpub response:', parseError, xpubResponse);
+          throw new Error(`Invalid JSON response from snap: ${parseError instanceof Error ? parseError.message : 'Unknown parse error'}`);
+        }
+      }
+
+      const newXpub = (parsedResponse as { response?: { xpub?: string } })?.response?.xpub;
+
+      if (!newXpub) {
+        throw new Error('Failed to get xpub from snap - no xpub in response');
+      }
+
+      setLoadingStep('Initializing read-only wallet...');
+
+      const newNetwork = DEFAULT_NETWORK;
+      await readOnlyWalletService.initialize(newXpub, newNetwork);
+
+      setupEventListeners();
+
+      setLoadingStep('Loading wallet data...');
+
+      let walletAddress = '';
+      try {
+        walletAddress = await getDisplayAddressForMode(addressMode, readOnlyWalletService);
+      } catch {
+        throw new Error('Failed to retrieve wallet address. The wallet may not be properly initialized.');
+      }
+
+      setLoadingStep('Loading balance...');
+
+      const walletBalances = await readOnlyWalletService.getBalance(TOKEN_IDS.HTR);
+
+      const genesisHash = '';
+      const tokenLoadResult = await loadTokensWithBalances(newNetwork, genesisHash, {
+        detailedErrors: false,
+      });
+
+      localStorage.setItem(STORAGE_KEYS.XPUB, newXpub);
+      localStorage.setItem(STORAGE_KEYS.NETWORK, newNetwork);
+
+      setIsConnected(true);
+      setAddress(walletAddress);
+      setBalances(walletBalances);
+      setNetwork(newNetwork);
+      setXpub(newXpub);
+      setIsConnecting(false);
+      setLoadingStep('');
+      setRegisteredTokens(tokenLoadResult.tokens);
+      onError(tokenLoadResult.warning);
+    } catch (error) {
+      if (error instanceof SnapUnauthorizedError) {
+        handleSnapError(error);
+        setIsConnecting(false);
+        setLoadingStep('');
+        return;
+      }
+
+      const errorMessage = error instanceof Error ? error.message : String(error);
+
+      const snapCrashed =
+        errorMessage.includes('DataCloneError') ||
+        errorMessage.includes('postMessage') ||
+        errorMessage.includes('cloned') ||
+        errorMessage.includes('timeout');
+
+      const userMessage = snapCrashed
+        ? 'MetaMask Snap is not responding. Please refresh the page and try again.'
+        : errorMessage || 'Failed to connect to wallet';
+
+      setIsConnecting(false);
+      setLoadingStep('');
+      onError(userMessage);
+    } finally {
+      setIsConnecting(false);
+      setLoadingStep('');
+    }
+  };
+
+  const disconnectWallet = () => {
+    const keysToRemove = Object.keys(localStorage).filter(key =>
+      key.startsWith('hathor')
+    );
+    keysToRemove.forEach(key => localStorage.removeItem(key));
+
+    readOnlyWalletService.stop();
```

Comment:
From what I remember we did not setup event listeners or websocket, so what is there to stop on the lib?
Would just deleting the object not clear everything?